source/common/network/connection_impl.cc

```diff
@@ -195,7 +195,7 @@ void ConnectionImpl::onRead(uint64_t read_buffer_size) {
     return;
   }
 
-  if (read_buffer_size == 0) {
+  if (read_buffer_size == 0 && !read_last_byte_) {
```

Comment:
I noticed a TODO below for handling half closed connections which we'll want to update.

That section also has to do with scheduling resumption in the case we have data to send. I'm 90% convinced we can never get a FIN without either having buffered data or sending it on, but _optionally_ I'd totally appreciate a sent_read_last_byte_ boolean we set true when we've passed read_last_byte_ to other layers and some asserts we don't send it twice and then we could add checks to get me to 100% sure.

Reply:
This is a bit tricky because a filter may StopIteration.  But let me see if I can make this work; I agree it would be good to have the extra checking.